### PR TITLE
feat(artifact): deploy model on push completion

### PR DESCRIPTION
**Pull request by jvallesm**
Because

- When a user pushes a model to the registry, we want to trigger its deployment.

This commit

- Adds minor fixes and some code polish to the registry plugin.
- Captures the [PUT Manifest](https://distribution.github.io/distribution/#put-manifest) call and triggers the model deployment via gRPC.

# 🗒️ Notes

This was tested against the unmerged branch https://github.com/instill-ai/model-backend/tree/containerized-support.

### CODE REVIEW

File: plugins/registry/handler.go [hunk 1]
```diff
@@ -199,13 +211,26 @@ func (rh *registryHandler) relay(ctx context.Context, p registryHandlerParams) {
...
+		// safe to repeat them if we fail here.
+		//
+		// TODO in the future the registry will handle more than model images,
+		// so this operation won't always be necessary. A much better pattern
+		// is publishing the push operation success as an event and let the
+		// clients to consume and act upon it (artifact to register the tag
+		// creation time, model to deploy the image...).
+
+		// TODO call UserModel or OrganizationModel depending on username == namespace.
+		modelName := fmt.Sprintf("users/%s/models/%s:%s", namespace, repository, resourceID)
```
Comment (jvallesm) [inline plugins/registry/handler.go:226]: @heiruwu @tonywang10101 here the model name would be something like:

```
users/jvalles/models/jvalles/llava-34b:latest
```

❓ Is that correct?

File: plugins/registry/handler.go [hunk 2]
```diff
@@ -199,13 +211,26 @@ func (rh *registryHandler) relay(ctx context.Context, p registryHandlerParams) {
...
+		// Deploy model.The previous operations are idempotent so it should be
+		// safe to repeat them if we fail here.
+		//
+		// TODO in the future the registry will handle more than model images,
+		// so this operation won't always be necessary. A much better pattern
+		// is publishing the push operation success as an event and let the
+		// clients to consume and act upon it (artifact to register the tag
+		// creation time, model to deploy the image...).
+
+		// TODO call UserModel or OrganizationModel depending on username == namespace.
```
Comment (jvallesm) [inline plugins/registry/handler.go:225]: @heiruwu @tonywang10101 it's a bit cumbersome to leak the user/org business domain here to call different endpoints 🙃 In VDP we discussed reducing the number of org/user duplicated endpoints, in most of them the underlying logic is the same. Not required for the immediate future but worth considering in the Model domain as well.

**Comment by jvallesm** [comment]
# QA ✅ 

### Login & push:
CleanShot 2024-03-22 at 19 16 53

### `api-gateway` (augmented) logs:
CleanShot 2024-03-22 at 19 16 06

### `model-backend` logs
CleanShot 2024-03-22 at 19 16 38

**Comment by jvallesm** [comment]
QA with the most recent contracts: ✅ 
CleanShot 2024-03-26 at 18 29 30

**Review by heiruwu — APPROVED** [review]

**Pull request merged**